### PR TITLE
fix(blame_line): Do not omit carriage return

### LINKED ISSUE #410: Gitsigns behavior is inconsistent in CRLF files
### Description

`git --version`: git version 2.34.0

When opening any file with windows line endings (`\r\n`) every line is reported as being changed, so that gitsigns is not useful anymore working on these buffers.

This behavior has appeared relatively recently (~ 2-4 weeks) but I haven't bisected that behavior yet. ~I can not fully rule out, that this has something to do with my git config, might also have to test with the default git config (is it possible to point to another git executable in `setup()`?)~ Behavior is the same, even if `~/.config/git/config` is deleted.

I think, this might have something to do, that somewhere on the line the checked in windows.txt file will be converted from CRLF to LF, while the actual file is still CRLF. This then leads to the diff between the two. 

#### Updates

- After some digging removing my custom git config and redoing the step again, the result is not slightly different:
  `windows.txt` lines are now reported as not committed yet, which is also wrong. I'll look further ...

- TBH I'm quite confused. When I remove the `require('packer').sync()` call in `minimal.lua` the issues disappears (did not change anything else) 
    - ... seems like it does not load the plugin, that's why.

- Aha, to even start blaming, this has to be done to `minimal.lua`

```diff
@@ -37,6 +37,4 @@ _G.load_config = function()
   vim.cmd('Gitsigns debug_messages')
 end

-require('packer').sync()
-
-vim.cmd [[autocmd User PackerComplete ++once echo "Ready!" | lua load_config()]]
+load_config()

```

- I've changed the `printf ...` examples to have newlines to circumvent issues related to `:help eol`. But the behavior is a little different again: 
  - `Gitsigns blame_line` on `unix.txt` behaves as expected
  -  `Gitsigns blame_line` on `windows.txt` reports "Not Committed Yet", even though the file is committed and no line is changed. 

- Another new finding: Using `:lua require('gitsigns').blame_line({ignore_whitespace=true})` instead for `windows.txt` fixes the behavior described the point above. 

### Neovim version

<details><summary><code>NVIM v0.6.0-dev+627-gb5b025f6a3</code></summary>
```
NVIM v0.6.0-dev+627-gb5b025f6a3 Build type: RelWithDebInfo LuaJIT 2.0.5 Compilation: /usr/sbin/cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions     -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security     -fstack-clash-protection -fcf-protection -Wp,-U_FORTIFY_SOURCE -Wp,-D_FORTIFY_SOURCE=1 -DNVIM_TS_HAS_SET_MATCH_LIMIT -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/violfa/.cache/paru/clone/neovim-git/src/build/config -I/home/violfa/.cache/paru/clone/neovim-git/src/neovim-git/src -I/usr/include -I/home/violfa/.cache/paru/clone/neovim-git/src/build/src/nvim/auto -I/home/violfa/.cache/paru/clone/neovim-git/src/build/include Compiled by violfa  Features: +acl +iconv +tui See ":help feature-compile"     system vimrc file: "$VIM/sysinit.vim"   fall-back for $VIM: "/usr/share/nvim"  Run :checkhealth for more info
```
</details>

### Operating system and version

WSL Arch Linux Kernel 5.10.60.1-microsoft-standard-WSL2

### Steps to reproduce

0. `mv ~/.config/git/config ~/.config/git/config.bak`
1. `mkdir gitsignsIssueRepro && cd gitsignsIssueRepro`
2. `git init && git config --local user.name "test" && git config --local user.email "test@test.test"`
3. `printf "This\nis\na\nunix\nfile\n" > unix.txt`
4. `printf "This\r\nis\r\na\r\nwindows\r\nfile\r\n" > windows.txt`
5. `git add unix.txt windows.txt`
6. `git commit -m 'Initial commit'`
7. `nvim -nu minimal.lua -o unix.txt windows.txt`

### Expected behavior

Both files should be reported as unchanged. (No signs)

### Actual behavior

`windows.txt` is reported as being changed (every line is reported as changed)

### Gitsigns debug messages

<details><summary><code>Gitsigns debug_messages</code></summary>

```
setup_highlight: Deriving GitSignsDelete from DiffDelete
setup_highlight: Highlight GitSignsDelete is already defined
setup_highlight: Deriving GitSignsAdd from DiffAdd
setup_highlight: Deriving GitSignsChange from DiffChange
setup_highlight: Highlight GitSignsChange is already defined
run_job: git --no-pager --version
attach(1): Attaching (trigger=setup)
run_job: git --no-pager config user.name
run_job: git --no-pager rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
run_job: git --no-pager --git-dir=/tmp/gitsignsIssueRepro/.git ls-files --stage --others --exclude-standard /tmp/gitsignsIssueRepro/unix.txt
watch_gitdir(1): Watching git dir
run_job: git --no-pager --git-dir=/tmp/gitsignsIssueRepro/.git show :0:unix.txt
update(1): updates: 1, jobs: 5
attach(2): Attaching (trigger=setup)
run_job: git --no-pager config user.name
run_job: git --no-pager rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
run_job: git --no-pager --git-dir=/tmp/gitsignsIssueRepro/.git ls-files --stage --others --exclude-standard /tmp/gitsignsIssueRepro/windows.txt
watch_gitdir(2): Watching git dir
run_job: git --no-pager --git-dir=/tmp/gitsignsIssueRepro/.git show :0:windows.txt
update(2): updates: 2, jobs: 9
run_job: git --no-pager --git-dir=/tmp/gitsignsIssueRepro/.git blame --contents - -L 1,+1 --line-porcelain /tmp/gitsignsIssueRepro/unix.txt
run_job: git --no-pager --git-dir=/tmp/gitsignsIssueRepro/.git blame --contents - -L 1,+1 --line-porcelain /tmp/gitsignsIssueRepro/windows.txt

```

</details>

<details><summary><code>:Gitsigns dump_cache</code></summary>

```
{ {
    compare_file = "/tmp/lua_5aa96s",
    compare_text = { "...",
      head = "This",
      length = 5
    },
    file = "/tmp/gitsignsIssueRepro/unix.txt",
    git_obj = {
      file = "/tmp/gitsignsIssueRepro/unix.txt",
      has_conflicts = false,
      mode_bits = "100644",
      object_name = "6bc13fc6628912423981e0f4206684e98986aa78",
      relpath = "unix.txt",
      repo = {
        abbrev_head = "master",
        gitdir = "/tmp/gitsignsIssueRepro/.git",
        toplevel = "/tmp/gitsignsIssueRepro",
        username = "test"
      }
    },
    gitdir_watcher = <userdata 1>,
    hunks = {},
    pending_signs = {},
    staged_diffs = {}
  }, {
    compare_file = "/tmp/lua_7CL1cF",
    compare_text = { "...",
      head = "This\r",
      length = 5
    },
    file = "/tmp/gitsignsIssueRepro/windows.txt",
    git_obj = {
      file = "/tmp/gitsignsIssueRepro/windows.txt",
      has_conflicts = false,
      mode_bits = "100644",
      object_name = "1544e0878be37ac55896285e645af381bfa7e213",
      relpath = "windows.txt",
      repo = {
        abbrev_head = "master",
        gitdir = "/tmp/gitsignsIssueRepro/.git",
        toplevel = "/tmp/gitsignsIssueRepro",
        username = "test"
      }
    },
    gitdir_watcher = <userdata 2>,
    hunks = {},
    pending_signs = {},
    staged_diffs = {}
  } }

```
</details>

###### Old

</details>

<details><summary><code>:Gitsigns dump_cache</code></summary>

```
{ {
    compare_file = "/tmp/lua_k0GksT",
    compare_text = { "...",
      head = "This",
      length = 5
    },
    file = "/tmp/gitsignsIssueRepro/unix.txt",
    git_obj = {
      file = "/tmp/gitsignsIssueRepro/unix.txt",
      has_conflicts = false,
      mode_bits = "100644",
      object_name = "b7d7480c838ebef9cd09902d08b96acad81f585a",
      relpath = "unix.txt",
      repo = {
        abbrev_head = "master",
        gitdir = "/tmp/gitsignsIssueRepro/.git",
        toplevel = "/tmp/gitsignsIssueRepro",
        username = "test <test@test.test>"
      }
    },
    gitdir_watcher = <userdata 1>,
    hunks = {},
    pending_signs = {},
    staged_diffs = {}
  }, {
    compare_file = "/tmp/lua_QQUW2Q",
    compare_text = { "...",
      head = "This\r",
      length = 5
    },
    file = "/tmp/gitsignsIssueRepro/windows.txt",
    git_obj = {
      file = "/tmp/gitsignsIssueRepro/windows.txt",
      has_conflicts = false,
      mode_bits = "100644",
      object_name = "685595e11442ead7e33041c190e821ec3828f2f6",
      relpath = "windows.txt",
      repo = {
        abbrev_head = "master",
        gitdir = "/tmp/gitsignsIssueRepro/.git",
        toplevel = "/tmp/gitsignsIssueRepro",
        username = "test <test@test.test>"
      }
    },
    gitdir_watcher = <userdata 2>,
    hunks = { {
        added = {
          count = 1,
          lines = { "file\r" },
          start = 5
        },
        dend = 5,
        head = "@@ -5,1 +5,1 @@",
        removed = {
          count = 1,
          lines = { "file" },
          start = 5
        },
        start = 5,
        type = "change",
        vend = 5
      } },
    pending_signs = {},
    staged_diffs = {}
  } }
```
</details>

<details><summary><code>:Gitsigns dump_cache</code></summary>

```
{ {
    compare_file = "/tmp/lua_7wAkU6",
    compare_text = { "...",
      head = "This",
      length = 5
    },
    file = "/tmp/gitsignsIssueRepro/unix.txt",
    git_obj = {
      file = "/tmp/gitsignsIssueRepro/unix.txt",
      has_conflicts = false,
      mode_bits = "100644",
      object_name = "b7d7480c838ebef9cd09902d08b96acad81f585a",
      relpath = "unix.txt",
      repo = {
        abbrev_head = "master",
        gitdir = "/tmp/gitsignsIssueRepro/.git",
        toplevel = "/tmp/gitsignsIssueRepro",
        username = "Fabian Viöl"
      }
    },
    gitdir_watcher = <userdata 1>,
    hunks = {},
    pending_signs = {},
    staged_diffs = {}
  }, {
    compare_file = "/tmp/lua_H9EMPB",
    compare_text = { "...",
      head = "This",
      length = 5
    },
    file = "/tmp/gitsignsIssueRepro/windows.txt",
    git_obj = {
      file = "/tmp/gitsignsIssueRepro/windows.txt",
      has_conflicts = false,
      mode_bits = "100644",
      object_name = "2bd326f77da19c32188c1be39b29e01e9641307e",
      relpath = "windows.txt",
      repo = {
        abbrev_head = "master",
        gitdir = "/tmp/gitsignsIssueRepro/.git",
        toplevel = "/tmp/gitsignsIssueRepro",
        username = "Fabian Viöl"
      }
    },
    gitdir_watcher = <userdata 2>,
    hunks = { {
        added = {
          count = 5,
          lines = { "This\r", "is\r", "a\r", "windows\r", "file\r" },
          start = 1
        },
        dend = 5,
        head = "@@ -1,5 +1,5 @@",
        removed = {
          count = 5,
          lines = { "This", "is", "a", "windows", "file" },
          start = 1
        },
        start = 1,
        type = "change",
        vend = 5
      } },
    pending_signs = {},
    staged_diffs = {}
  } }
```
</details>

### Minimal config

```Lua
vim.cmd [[set runtimepath=$VIMRUNTIME]]
vim.cmd [[set packpath=/tmp/nvim/site]]
local package_root = '/tmp/nvim/site/pack'
local install_path = package_root .. '/packer/start/packer.nvim'

if vim.fn.isdirectory(install_path) == 0 then
  print("Installing Gitsigns and dependencies.")
  -- note: No --depth 1 to get the full log for bisecting
  vim.fn.system { 'git', 'clone', 'https://github.com/wbthomason/packer.nvim', install_path }
end

-- Load plugins
require('packer').startup {
  {
    'wbthomason/packer.nvim',
    {
      'lewis6991/gitsigns.nvim',
      requires = { 'nvim-lua/plenary.nvim' },
    },
    -- ADD PLUGINS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
  },
  config = {
    package_root = package_root,
    compile_path = install_path .. '/plugin/packer_compiled.lua',
    display = { non_interactive = true },
  },
}

_G.load_config = function()
  require('gitsigns').setup{
    debug_mode = true, -- You must add this to enable debug messages
    -- GITSIGNS CONFIG THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
  }

  -- ADD INIT.LUA SETTINGS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE

  -- ALSO INCLUDE GITSIGNS DEBUG MESSAGES
  vim.cmd('Gitsigns debug_messages')
end

require('packer').sync()

vim.cmd [[autocmd User PackerComplete ++once echo "Ready!" | lua load_config()]]
```

### PR DESCRIPTION
**Pull request by Sh3Rm4n**
When blaming a line on a file, which as the "dos" fileformat
which is also checked in as a "dos" format including the
carriage returns, gitsigns.nvim currently reports
"Not yet commited" caused by an unsucessfull blame call
to git.

This commit fixes that behavior by appending the carriage
return lines again

- [x] Have you read [CONTRIBUTING.md](https://github.com/lewis6991/gitsigns.nvim/blob/HEAD/CONTRIBUTING.md)?

Fixes #410 

Starting as a draft, because the tests do not really make sense. Do you have a suggestion, which test helper function to use, to actually test the content of `blame_line()` @lewis6991?

### CODE REVIEW

File: teal/gitsigns/actions.tl [hunk 1]
```diff
@@ -399,7 +399,8 @@ M.preview_hunk = noautocmd(function()
   if not hunk then return end
 
   local hlines = gs_hunks.patch_lines(hunk)
-  if vim.bo[cbuf].fileformat == 'dos' then
+  -- TODO: Decide on this
+  if vim.api.nvim_buf_get_option(cbuf, 'fileformat') == 'dos' then
```
Comment (lewis6991) [inline teal/gitsigns/actions.tl:403]: Unless there is a specific reason. vim.* is prefered over vim.api.*

File: test/gitsigns_spec.lua [hunk 2]
```diff
@@ -256,6 +256,19 @@ describe('gitsigns', function()
...
+      setup_test_repo(true)
+      config.current_line_blame = true
+      setup_gitsigns(config)
+      edit(newfile)
+      command("set fileformat=dos")
+      -- TODO: What does insert do?
+      insert({"This\r", "is\r", "a\r", "dos\r", "file\r"})
+      command("write")
+      git{'add', newfile}
+      git{"commit", "-m", "commit on main"}
```
Comment (lewis6991) [inline test/gitsigns_spec.lua:270]: Since we're using current_line_blame, we can use Screen to test this. Take a look at the 'can enable numhl' test.

**Comment by Sh3Rm4n** [comment]
The issue I've described in #410 was not fixed for me with #417, so I finally came around and updated this PR.  Though, the description of #410 was a little confusing TBH, because I wrote the issue while still exploring the real origin of the problem.

The essential fix was to move the existing `buf_lines` function to the `util` module and use it instead of `vim.api.nvim_buf_get_lines`

There are multiple code paths which are effected (signs, `blame_line` and `currrent_line_blame`) but the tests do only test for the `current_line_blame` behavior for now.

**Comment by Sh3Rm4n** [comment]
Not sure how to go about the [`screen`](https://github.com/lewis6991/gitsigns.nvim/runs/4498775064?check_suite_focus=true#step:10:59) warning, though.

**Comment by lewis6991** [comment]
No worries, I can take a look when I get some time.

File: Makefile [hunk 3]
```diff
@@ -1,7 +1,9 @@
 
 export PJ_ROOT=$(PWD)
 
-FILTER=.*
+ifeq (,$(FILTER))
+FILTER := .*
+endif
```
Comment (lewis6991) [inline Makefile:6]: Why is this needed?

File: teal/gitsigns/actions.tl [hunk 4]
```diff
@@ -788,15 +788,13 @@ M.diffthis = void(function(base: string)
...
-  local ff = vim.bo[bufnr].fileformat
-
   local text: {string}
   local err: string
   local comp_rev = bcache:get_compare_rev(calc_base(base))
 
   if base then
     text, err = bcache.git_obj:get_show_text(comp_rev)
-    if ff == 'dos' then
+    if vim.bo[bufnr].fileformat == 'dos' then
```
Comment (lewis6991) [inline teal/gitsigns/actions.tl:797]: This will error since this line runs in a fast event as `get_show_text()` is async. You can check this with `:Gitsigns diffthis ~`.

File: Makefile (hunk 3 above)
Comment (Sh3Rm4n) [reply Makefile:6]: This allows to overwrite the FILTER variable, to run specific tests, meaning if FILTER is not set, fall back to `.*` as the default. 

For example, I used it to run only my tests, via:
```
FILTER='.*current line blame does ' make test
```

Comment (lewis6991) [reply Makefile:6]: But if FILTER is unset then that is the same as `.*`

Comment (Sh3Rm4n) [reply Makefile:6]: You mean, `--filter=''` is the same as `--filter='.*'`?

Comment (lewis6991) [reply Makefile:6]: Ok I see what you're trying to do. You needed to quote FILTER on line 40, it no longer works if it's empty.

In that case you do can: 
```
FILTER ?= .*
```

Comment (Sh3Rm4n) [reply Makefile:6]: Ah, thanks. I always forget that this assignment type exists. Will fix.

**Comment by lewis6991** [comment]
Don't worry about the warning for now.

Thanks for the patch :+1:

If you have bandwidth, I'd greatly appreciate help with any windows specific issues in the future 🤞 .
